hopr-socks/src/main.rs

```diff
@@ -0,0 +1,203 @@
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let opt: Opt = Opt::from_args();
+    let socks_domain: String = [opt.host, opt.port].join(":");
+
+    env_logger::init();
+
+    return match opt.mode {
+        RunModeOpt::Client {
+            target_host,
+            target_port,
+            auth,
+        } => spawn_socks_client(&socks_domain, target_host, target_port, auth).await,
+        RunModeOpt::Server { request_timeout, auth } => spawn_socks_server(&socks_domain, request_timeout, auth).await,
+    };
+}
+
+async fn spawn_socks_client(
+    socks_domain: &String,
+    target_host: String,
+    target_port: u16,
+    auth: AuthMode,
+) -> Result<()> {
+    let remote_domain: String = [target_host.clone(), target_port.to_string()].join(":");
+    let config = ClientConfig::default();
+
+    // Creating a SOCKS stream to the target address through the socks server
+    let mut socks = match auth {
+        AuthMode::NoAuth => Socks5Stream::connect(socks_domain.clone(), target_host, target_port, config).await?,
+        AuthMode::Password { username, password } => {
+            Socks5Stream::connect_with_password(
+                socks_domain.clone(),
+                target_host,
+                target_port,
+                username,
+                password,
+                config,
+            )
+            .await?
+        }
+    };
+
+    // Once connection is completed, can start to communicate with the server
+    http_request(&mut socks, remote_domain).await?;
+
+    Ok(())
+}
+async fn spawn_socks_server(socks_domain: &String, timeout: u64, auth: AuthMode) -> Result<()> {
+    let mut config = ServerConfig::default();
+    config.set_request_timeout(timeout);
+
+    let config = match auth {
+        AuthMode::NoAuth => {
+            warn!("No authentication has been set!");
+            config
+        }
+        AuthMode::Password { username, password } => {
+            info!("Simple auth system has been set.");
+            config.with_authentication(SimpleUserPassword { username, password })
+        }
+    };
+
+    let config = Arc::new(config);
+    let listener = TcpListener::bind(&socks_domain).await?;
+
+    info!("Listen for socks connections @ {}", &socks_domain);
+
+    // Standard TCP loop
+    loop {
+        match listener.accept().await {
+            Ok((socket, _addr)) => {
+                info!("Connection from {}", socket.peer_addr()?);
+                let socket = Socks5Socket::new(socket, config.clone());
+
+                spawn_and_log_error(socket.upgrade_to_socks5());
+            }
+            Err(err) => error!("accept error = {:?}", err),
+        }
+    }
+}
+
+fn spawn_and_log_error<F, T, A>(fut: F) -> task::JoinHandle<()>
+where
+    F: Future<Output = Result<Socks5Socket<T, A>>> + Send + 'static,
+    T: AsyncRead + AsyncWrite + Unpin,
+    A: Authentication,
+{
+    task::spawn(async move {
+        if let Err(e) = fut.await {
+            error!("{:#}", &e);
+        }
```

Comment:
This should not be needed though, as `spawn` returns a `JoinHandle<_>`, so you could do:

```rust
if let Err(e) = task::spawn(fut).await {
    error!("{e}")
}

Reply:
Can make it work without. I'll leave it as it for now.